addons/product/populate/product.py

```diff
@@ -70,11 +71,26 @@ def get_rand_float(values, counter, random):
         ]
 
     def _populate_factories(self):
+        # Generate a unique `barcode` prefix so consecutive populate runs don't
+        # raise a unique constraint error on `barcode`.
+        barcode_prefix = 'BARCODE-PP-'
+        query = SQL("""
+            SELECT barcode
+              FROM product_product
+             WHERE barcode LIKE %s
+          ORDER BY id DESC
+             LIMIT 1
+        """, barcode_prefix + '%')
+        self.env.cr.execute(query)
+        barcode = self.env.cr.fetchone()
+        if barcode:
+            barcode_prefix = barcode[0] + '-'
+
         return [
             ("name", populate.constant('product_product_name_{counter}')),
             ("description", populate.constant('product_product_description_{counter}')),
             ("default_code", populate.constant('PP-{counter}')),
-            ("barcode", populate.constant('BARCODE-PP-{counter}')),
+            ("barcode", populate.constant(barcode_prefix + '{counter}')),
```

Comment:
what about removing this line entirely, and allowing https://github.com/odoo/odoo/blob/17.0/addons/product/populate/product_template.py#L66 to set the barcodes on variants instead?

Reply:
That code is not related to the issue and would not fix it as it's not called at all since we are fixing the populate of `product.product`, not `product.template` (which is what `account` module is using as `_populate_dependencies`).
The fix has to basically make this to not crash:
```bash
<odoo-bin-command-and-addons> populate -d populate --size=small --models=product.product
<odoo-bin-command-and-addons> populate -d populate --size=small --models=product.product
```

Reply:
My bad, I didn't think of `--models=` and assumed that we'd always get `product.product`s with barcodes anyway.

Anyway, IMO the way that `product.template` sets barcodes on some of it's variants is cleaner and wouldn't lead to longer barcodes on every populate.
From an accounting perspective, we don't really need barcodes, so I leave it up to you guys and the codeowners of `product`

However, I'll just leave my 2 cents and suggest removing `barcode` from the list returned by `_populate_factories` and instead add the barcodes after

```py
def _populate(self, size):
    res = super()._populate(size)

    for product in res:
        product.barcode = "BARCODE-PP-%s" % product.id

    return res
```

Reply:
Writing in a loop, I'd avoid that.
That would make the SQL queries skyrocket, basically 3 extra SQL Queries **per** created product.
While the current solution does not imply any extra sql query.

I checked out of curiosity:

|  | current solution | your solution |
| - | - | - |
| **SELECT Queries** | | |
|  product.product | 0 | 151 |
| product.packaging | 0 | 151 |
| **UPDATE  Queries**|  | |
|  product.product | 1 | 152 |

The populate command being already very slow, I'd avoid that.